api/submit_result.go

```diff
@@ -23,21 +27,49 @@ func newResultSubmitter(api *API) *resultSubmitter {
 }
 
 // Submit submits results for executionID.
-func (s *resultSubmitter) Submit(executionID string, outputKey string, outputData map[string]interface{}) error {
-	exec, err := s.api.execDB.Find(executionID)
+func (s *resultSubmitter) Submit(executionID string, outputKey string, outputData []byte) error {
+	exec, stateChanged, err := s.processExecution(executionID, outputKey, outputData)
+	if stateChanged {
+		// only publish to listeners when the execution's state changed.
+		go pubsub.Publish(exec.Service.ResultSubscriptionChannel(), exec)
+	}
+	// always return any error to the service.
+	return err
+}
+
+// processExecution processes execution and marks it as complated or failed.
+func (s *resultSubmitter) processExecution(executionID string, outputKey string, outputData []byte) (exec *execution.Execution, stateChanged bool, err error) {
+	stateChanged = false
+	exec, err = s.api.execDB.Find(executionID)
 	if err != nil {
-		return err
+		return nil, false, err
 	}
+
 	exec.Service, err = service.FromService(exec.Service, service.ContainerOption(s.api.container))
 	if err != nil {
-		return err
+		return s.saveExecution(exec, err)
+	}
+
+	var outputDataMap map[string]interface{}
+	if err := json.Unmarshal(outputData, &outputDataMap); err != nil {
+		return s.saveExecution(exec, fmt.Errorf("invalid output data error: %s", err))
 	}
-	if err := exec.Complete(outputKey, outputData); err != nil {
-		return err
+
+	if err := exec.Complete(outputKey, outputDataMap); err != nil {
```

Comment:
Not for this PR, but we should also save the failed messages.

Reply:
can you elaborate, i don't understand

Reply:
An invalid json response we should also be kept inside db. Someone may want to see the task response (even if it's invalid). I'm even thinking about passing this response, not just an error. But it just a thought and needs lot way thinking if this approach is valid.